haskell/test/SMT/Model/BaseDAO/Types.hs

```diff
@@ -0,0 +1,269 @@
+                    Just (OtherContractType oc) ->
+                      let updatedSc = oc
+                            { ocStorage = param : (oc & ocStorage)
+                            , ocMutez = unsafeAddMutez tez (oc & ocMutez)
+                            }
+                      in M.insert addr (OtherContractType updatedSc) contractMap
+                    _ ->
+                      contractMap
+          )
+          (contracts)
+          (ops)
+  modify $ \ms -> ms { msContracts = updatedContracts }
+
+getStore :: ModelT Storage
+getStore = get <&> msFullStorage <&> fsStorage
+
+getConfig :: ModelT Config
+getConfig = get <&> msFullStorage <&> fsConfig
+
+modifyStore :: (Storage -> Storage) -> ModelT ()
+modifyStore f = do
+  ms <- get
+  put $ ms
+    { msFullStorage = (ms & msFullStorage)
+        { fsStorage = f (ms & msFullStorage & fsStorage) }
+    }
+
+isFA2Contract :: Address -> ModelT Bool
+isFA2Contract addr = do
+  ms <- get
+  case M.lookup addr (ms & msContracts) of
+    Just (SimpleFA2ContractType _) -> pure True
+    _ -> pure False
+
+mkSimpleOperation :: Operation -> SimpleOperation
+mkSimpleOperation = \case
+  (OpTransferTokens (TransferTokens param amt (VContract addr _))) ->
+    case cast param of
+      Just param' -> do
+        FA2TransferOperation addr (fromVal @[FA2.TransferItem] param') amt
+      Nothing ->
+        OtherOperation addr (show param) amt
+  op ->
+    error $ "Unable to convert `Operation` to `SimpleOperatoin`" <> show op
+
+-- | Version of `-$?` used in `ModelT`
+infixr 2 -$$?
+(-$$?) f x = case (f -$? x) of
+  Right a -> pure a
+  Left err -> throwError $ parseMichelsonFailed err
+
+-- Simple operation that only allows transfering to `FA2.TransferItem` entrypoint
+data SimpleOperation
+  = FA2TransferOperation Address [FA2.TransferItem] Mutez
+  | OtherOperation Address Text Mutez
+  deriving stock (Eq, Show, Generic)
+
+instance Buildable SimpleOperation where
+  build = genericF
+
+
+-- A simple contract type used to track originate contracts
+data SimpleContractType
+  = SimpleFA2ContractType SimpleFA2Contract
+  | OtherContractType OtherContract
+  deriving stock (Eq, Show, Generic)
+
+instance Buildable SimpleContractType where
+  build = genericF
+
+data SimpleFA2Contract = SimpleFA2Contract
+  { sfcStorage :: [FA2.TransferParams]
+  , sfcMutez :: Mutez
+  } deriving stock (Eq, Show, Generic)
+
+instance Buildable SimpleFA2Contract where
+  build = genericF
+
+
+data OtherContract = OtherContract
+  { ocStorage :: [Text]
+  , ocMutez :: Mutez
+  } deriving stock (Eq, Show, Generic)
+
+instance Buildable OtherContract where
+  build = genericF
+
+-- | Definition of an entrypoint call to emulate in the SMTs
+data ModelCall = ModelCall
+  { mcSender :: Address
+  , mcAdvanceLevel :: Maybe Natural -- Advance level before calling the entrypoint
+  , mcParameter :: Parameter
+  , mcSource :: Address
+  } deriving stock (Eq, Show, Generic)
```

Comment:
Given that there is no use for "advance level" and "parameter" after the very beginning, I suggest splitting the two "sources" of the call in a separate type and pass that to the `apply*` functions only.
```suggestion
-- | Definition of an entrypoint call to emulate in the SMTs
data ModelCall = ModelCall
  { mcAdvanceLevel :: Maybe Natural -- Advance level before calling the entrypoint
  , mcParameter :: Parameter
  , mcSources :: ModelSource
  } deriving stock (Eq, Show, Generic)

-- | Definition of an @source@ and @sender@ for a call to emulate in the SMTs
data ModelSource = ModelSource
  { msSender :: Address
  , msSource :: Address
  } deriving stock (Eq, Show, Generic)
```

Reply:
Updated.